### PR TITLE
Expose public edge-harvesting HNSW insertion API and docs

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Expose a public edge-harvesting HNSW insertion API and document its execution plan and roadmap completion. The change set now includes a new public insertion API wrapper, updated design/roadmap docs, a dedicated ExecPlan, expanded tests, and governance updates.

### New Features
- Add CpuHnsw::insert_harvesting to allow callers to insert points and receive harvested candidate edges.

### Enhancements
- Update design documentation and roadmap to reflect the new public edge-harvesting insertion path and its completion.
- Refine Markdown tables, escaping, and references across design and roadmap docs for consistency and formatting correctness.

### Documentation
- Introduce an execution plan document for roadmap item 11.1.1 describing constraints, risks, plan of work, and acceptance criteria for the edge-harvesting API change.
- Update docs/chutoro-design.md with formatting and wording adjustments reflecting the new API surface and governance context.

### Tests
- Add unit tests validating insert_harvesting behaviour, including initial insert, valid edge invariants, duplicate insertion handling, and graph-state parity with the existing insert path.
- Extend test coverage to exercise edge harvesting under multiple insertion scenarios and parity checks against the standard insert path.

### ExecPlan
- Added docs/execplans/11-1-1-make-edge-harvesting-hnsw-insertion-path-public.md describing the plan, risks, and validation steps.

### Roadmap
- Updated docs/roadmap.md to mark 11.1.1 as completed (x) and reflect the new public API.

### Artifacts (Git changes)
- chutoro-core/src/hnsw/cpu/mod.rs
- chutoro-core/src/hnsw/tests/edge_harvest.rs
- docs/chutoro-design.md
- docs/execplans/11-1-1-make-edge-harvesting-hnsw-insertion-path-public.md
- docs/roadmap.md

## Notes
- The public insert_harvesting API is implemented as a thin wrapper over the existing insert_with_edges path, preserving existing insert() behavior while enabling external callers to obtain harvested edges for incremental clustering workflows.
- Tests cover bootstrap (initial insert), edge invariants, duplicate handling, and graph-state parity with insert().
- The ExecPlan document and design/roadmap updates provide governance and validation guidance for roadmap item 11.1.1.

## Validation plan (high level)
- Build and run unit tests for chutoro-core HNSW edge harvesting tests.
- Run fmt, lint, and markdown validators.
- Verify that external callers can use CpuHnsw::insert_harvesting and receive Vec<CandidateEdge>.

📎 **Task**: https://www.devboxer.com/task/c4989484-236f-41c5-bb29-f7dbeacfa51b

### CODE REVIEW

File: docs/roadmap.md [hunk 1]
```diff
@@ -1202,18 +1202,18 @@ Provenance notes:
...
-      <https://scikit-learn.org/stable/datasets/real_world.html#newsgroups-dataset>
-[^6]: RCV1-v2 —
-      <https://scikit-learn.org/stable/modules/generated/sklearn.datasets.fetch_rcv1.html>
-[^7]: SNAP datasets — <https://snap.stanford.edu/data/>
-[^8]: PBMC 68k —
-      <https://support.10xgenomics.com/single-cell-gene-expression/datasets>
-[^9]: GloVe vectors — <https://nlp.stanford.edu/projects/glove/>
-[^10]: ANN-Benchmarks datasets —
-       <https://github.com/erikbern/ann-benchmarks>
+\[^1\]: scikit-learn — make_blobs —
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/roadmap.md]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

[the remark this answers is not in the document]
Comment (leynos) [reply docs/roadmap.md]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

Annul any requirements that violate the en-GB-oxendict spelling (-ize / -yse / -our) conventions (for example a request to replace "normalize" with "normalise" or "artefact" with "artifact"), or where the requirement unnecessarily increases cyclomatic complexity.

~~~
Please address the comments from this code review:

## Overall Comments
- In `CpuHnsw::insert_harvesting`’s Rustdoc example you call `CpuHnsw::build(&data, params)` and then `insert_harvesting(1, &data)`, which would actually be a duplicate insert for node 1 on a built index; consider changing the example to start from `with_capacity` and insert previously unseen nodes so it reflects real behaviour.
- The ExecPlan in `docs/execplans/11-1-1-make-edge-harvesting-hnsw-insertion-path-public.md` has `Status: COMPLETE` and a filled `Progress` section, but the `Outcomes & Retrospective` section still says work has not started yet; updating that narrative to match the current state would avoid future confusion.

## Individual Comments

### Comment 1
<location path="chutoro-core/src/hnsw/cpu/mod.rs" line_range="346" />
<code_context>
+    /// let index = CpuHnsw::build(&data, params).expect("build must succeed");
+    ///
+    /// // Insert a new node and harvest candidate edges
+    /// let edges = index.insert_harvesting(1, &data).expect("insert must succeed");
+    /// assert!(!edges.is_empty(), "inserted node should have candidate edges");
+    /// ```
</code_context>
<issue_to_address>
**issue (bug_risk):** The example inserts a node index that is likely already present in the built index, conflicting with the documented DuplicateNode behavior.

Given that `CpuHnsw::build(&data, params)` will usually index all items in `data`, node `1` is already present. `insert_harvesting(1, &data)` would then trigger `DuplicateNode`, contradicting both the docs and the `
</issue_to_address>

### Comment 2
<location path="docs/execplans/11-1-1-make-edge-harvesting-hnsw-insertion-path-public.md" line_range="149" />
<code_context>
+
+## Outcomes & Retrospective
+
+Work has not started yet. The intended completed state is a minimal
+public API addition with broad automated coverage, updated design notes,
+and all required validators passing without widening the task into
</code_context>
<issue_to_address>
**issue:** Outcomes section claims work has not started yet, which conflicts with the COMPLETE status and populated progress list.

Please update this paragraph to describe the actual completed outcome and any retrospective notes, so it’s clear this ExecPlan reflects finished work rather than planned work.
</issue_to_address>

### Comment 3
<location path="docs/execplans/11-1-1-make-edge-harvesting-hnsw-insertion-path-public.md" line_range="1" />
<code_context>
+# Expose a public edge-harvesting HNSW insertion API
+
+This ExecPlan (execution plan) is a living document. The sections
</code_context>
<issue_to_address>
**issue (review_instructions):** The acronym “HNSW” is used in the heading without being expanded on first use, which violates the requirement to define uncommon acronyms.

Consider expanding HNSW the first time it appears in this document, for example:

"# Expose a public edge-harvesting Hierarchical Navigable Small World (HNSW) insertion API"

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>
~~~

**Comment by leynos** [comment]
@coderabbitai please create a new GitHub issue for the following:

In `@docs/chutoro-design.md` around lines 966 - 968, The phrase "stable Rust
primitives with toolchain `1.93.1` (latest stable, released 2026-02-12)" is a
moving-target claim; change it to a time-scoped statement such as "stable as of
2026-03-02" or update the toolchain and date to the current stable release
(e.g., replace `1.93.1` and the parenthetical with the current stable version
and its release date). Locate the exact text in chutoro-design.md that reads
"stable Rust primitives with toolchain `1.93.1` (latest stable, released
2026-02-12)" and replace the parenthetical to either "stable as of 2026-03-02"
or the updated version/date to avoid the "latest stable" claim.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

### ❌ Failed checks (2 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| User-Facing Documentation | ⚠️ Warning | CpuHnsw::insert\_harvesting() is a public, exported API yet lacks documentation in the users-guide despite its stated scope of covering all exported APIs. | Add a section to docs/users-guide.md documenting the CpuHnsw API and insert\_harvesting() method with usage patterns and error handling information. |
| Developer Documentation | ⚠️ Warning | The new CpuHnsw::insert\_harvesting() public API lacks documentation in docs/developers-guide.md, though it is well-documented in Rustdoc, design documentation and ExecPlans. | Add a section to docs/developers-guide.md documenting core HNSW insertion methods, their signatures, return types, usage guidance and links to detailed design documentation. |

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai review
